lib/service/service.go

```diff
@@ -3239,6 +3247,124 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	return nil
 }
 
+// initDebugService starts debug service serving endpoints used for
+// troubleshooting the instance.
+func (process *TeleportProcess) initDebugService() error {
+	logger := process.logger.With(teleport.ComponentKey, teleport.Component(teleport.ComponentDebug, process.id))
+
+	// logPProfMiddleware logs /debug/pprof requests.
+	logPProfMiddleware := func(next http.HandlerFunc) http.HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request) {
+			seconds := r.URL.Query().Get("seconds")
+			if seconds == "" {
+				seconds = "default"
+			}
+
+			logger.InfoContext(
+				process.ExitContext(),
+				"Collecting pprof profile.",
+				"profile", strings.TrimSuffix(r.URL.Path, constants.PProfEndpointsPrefix),
+				"seconds", seconds,
+			)
+
+			next(w, r)
+		}
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc(constants.PProfEndpointsPrefix, logPProfMiddleware(pprof.Index))
+	mux.HandleFunc(constants.PProfEndpointsPrefix+"cmdline", logPProfMiddleware(pprof.Cmdline))
+	mux.HandleFunc(constants.PProfEndpointsPrefix+"profile", logPProfMiddleware(pprof.Profile))
+	mux.HandleFunc(constants.PProfEndpointsPrefix+"symbol", logPProfMiddleware(pprof.Symbol))
+	mux.HandleFunc(constants.PProfEndpointsPrefix+"trace", logPProfMiddleware(pprof.Trace))
+
+	mux.HandleFunc(constants.DebugServiceLogLevelEndpoint, func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == constants.DebugServiceGetLogLevelMethod {
+			switch process.Config.LoggerLevel.Level() {
+			case logutils.TraceLevel:
+				w.Write([]byte(logutils.TraceLevelString))
+			default:
+				w.Write([]byte(process.Config.LoggerLevel.Level().String()))
+			}
+
+			return
+		}
+
+		if r.Method != constants.DebugServiceSetLogLevelMethod {
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+		rawLevel, err := io.ReadAll(io.LimitReader(r.Body, 1024))
+		defer r.Body.Close()
+		if err != nil {
+			w.WriteHeader(http.StatusUnprocessableEntity)
+			w.Write([]byte("Unable to read request body."))
+			return
+		}
+
+		var level slog.Level
+
+		// TRACE string representaiton is not handled by slog. We need to
+		// convert it. Otherwise, parse the provided level.
+		if strings.EqualFold(string(rawLevel), logutils.TraceLevelString) {
+			level = logutils.TraceLevel
+		} else {
+			if err := level.UnmarshalText(rawLevel); err != nil {
+				w.WriteHeader(http.StatusUnprocessableEntity)
+				w.Write([]byte("Invalid log level."))
+				return
+			}
+		}
+
+		currLevel := process.Config.LoggerLevel.Level()
+		message := fmt.Sprintf("Log level already set to %q.", level)
+		if level != currLevel {
+			message = fmt.Sprintf("Changed log level from %q to %q.", currLevel, level)
```

Comment:
nit: when setting to trace, I see `Changed log level from "INFO" to "DEBUG-1".`. It's understandable though.